### PR TITLE
Simplify and speed up MemoryExec insert 

### PR DESCRIPTION
**Pull request by alamb**
~Draft until https://github.com/apache/arrow-datafusion/pull/6154 is merged~

# Which issue does this PR close?
Follow on to  https://github.com/apache/arrow-datafusion/pull/6154

# Rationale for this change
We had substantial back and forth on https://github.com/apache/arrow-datafusion/pull/6154 related to locking or a single batch on each insert or for the entire operation. This required two similar, but not quite the same codepaths

Upon reflection, since everything is being buffered anyways, we can avoid a second mostly copy of the loop and instead just insert the batches once at the end of the stream.

# What changes are included in this PR?

1. Pull state into a `struct` with names
2. Buffer batches in a separate `Vec` and then insert them all at once when the input stream is done
3. Remove the second mostly copy

# Are these changes tested?
covered by existing tests 

# Are there any user-facing changes?

### CODE REVIEW

File: datafusion/core/src/physical_plan/memory.rs [hunk 1]
```diff
@@ -327,43 +327,29 @@ impl ExecutionPlan for MemoryWriteExec {
...
-                as SendableRecordBatchStream
+        let state = StreamState {
+            data,
+            buffer: vec![],
+            batch: self.batches[partition % batch_count].clone(),
         };
+
+        let stream = futures::stream::unfold(state, |mut state| async move {
+            loop {
+                match state.data.next().await {
```
Comment (alamb) [inline datafusion/core/src/physical_plan/memory.rs:338]: This version of the loop does not hold the lock at all until the end.

**Comment by alamb** [comment]
@metesynnada  -- I wonder if you have some time to consider this change?

File: datafusion/core/src/physical_plan/memory.rs [hunk 2]
```diff
@@ -327,43 +327,29 @@ impl ExecutionPlan for MemoryWriteExec {
...
+                    Some(Ok(batch)) => state.buffer.push(batch),
+                    Some(Err(e)) => return Some((Err(e), state)),
+                    None => {
+                        // stream is done, transfer all data to target PartitionData
+                        state.batch.write().await.append(&mut state.buffer);
```
Comment (metesynnada) [inline datafusion/core/src/physical_plan/memory.rs:343]: I think this is logical, every stream will append once no matter the partition counts.

**Comment by metesynnada** [comment]
LGTM, @alamb. I truly appreciate your help and effort – thank you.

**Pull request merged**